File: data/api/src/main/java/io/github/droidkaigi/feeder/data/Authenticator.kt
package io.github.droidkaigi.feeder.data

import dev.gitlive.firebase.Firebase
import dev.gitlive.firebase.auth.auth
import io.github.droidkaigi.feeder.Authenticator
import io.github.droidkaigi.feeder.Logger
import io.github.droidkaigi.feeder.User
import javax.inject.Inject

class AuthenticatorImpl @Inject constructor(): Authenticator {

    override suspend fun currentUser() : User? {
        val firebaseUser = Firebase.auth.currentUser ?: return null
        val idToken = firebaseUser.getIdToken(false)

        return User(idToken)
    }

    override suspend fun signInAnonymously() : User? {
        val result = Firebase.auth.signInAnonymously()
        Logger.d("signin:${result.user}")

        val firebaseUser = result.user ?: return null
        val idToken = firebaseUser.getIdToken(false)

        return User(idToken)
    }

}

File: ios/Sources/Repository/Authenticator.swift
import DroidKaigiMPP
import FirebaseAuth
import Foundation

public class AuthenticatorImpl: Authenticator {

    public func currentUser(completionHandler: @escaping (DroidKaigiMPP.User?, Error?) -> Void) {
        guard let firebaseUser = Auth.auth().currentUser else {
            completionHandler(nil, nil)
            return
        }
        firebaseUser.getIDTokenResult { (result, error) in
            if let error = error {
                completionHandler(nil, error)
                return
            }
            guard let token = result?.token else {
                completionHandler(nil, nil)
                return
            }

            completionHandler(DroidKaigiMPP.User(idToken: token), nil)
        }
    }

    public func signInAnonymously(completionHandler: @escaping (DroidKaigiMPP.User?, Error?) -> Void) {
        Auth.auth().signInAnonymously { (result, error) in
            if let error = error {
                completionHandler(nil, error)
                return
            }

            guard let result = result else {
                completionHandler(nil, nil)
                return
            }

            result.user.getIDTokenResult { (result, error) in
                if let error = error {
                    completionHandler(nil, error)
                    return
                }
                guard let token = result?.token else {
                    completionHandler(nil, nil)
                    return
                }

                completionHandler(DroidKaigiMPP.User(idToken: token), nil)
            }
        }
    }

}
